velox/functions/sparksql/String.h

```diff
@@ -1431,4 +1432,208 @@ struct LevenshteinDistanceFunction {
+      const arg_type<Varchar>* upperCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        StringView{kMaskedLowerCase_},
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        StringView{kMaskedDigit_},
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        std::nullopt);
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>* inputPtr,
+      const arg_type<Varchar>* upperCharPtr,
+      const arg_type<Varchar>* lowerCharPtr,
+      const arg_type<Varchar>* digitCharPtr,
+      const arg_type<Varchar>* otherCharPtr) {
+    if (inputPtr == nullptr) {
+      return false;
+    }
+
+    doCall(
+        result,
+        *inputPtr,
+        getMaskedChar(upperCharPtr),
+        getMaskedChar(lowerCharPtr),
+        getMaskedChar(digitCharPtr),
+        getMaskedChar(otherCharPtr));
+    return true;
+  }
+
+ private:
+  void doCall(
+      out_type<Varchar>& result,
+      StringView input,
+      const std::optional<StringView> upperChar,
+      const std::optional<StringView> lowerChar,
+      const std::optional<StringView> digitChar,
+      const std::optional<StringView> otherChar) const {
+    auto inputBuffer = input.data();
+    const size_t inputSize = input.size();
+    result.reserve(inputSize);
+    auto outputBuffer = result.data();
+    size_t inputIdx = 0;
+    size_t outputIdx = 0;
+    while (inputIdx < inputSize) {
+      utf8proc_int32_t curCodePoint;
+      int charByteSize;
+      curCodePoint = utf8proc_codepoint(
+          &inputBuffer[inputIdx], inputBuffer + inputSize, charByteSize);
+      charByteSize = curCodePoint == -1 ? 1 : charByteSize;
+      auto maskedChar = &inputBuffer[inputIdx];
+      auto maskedCharByteSize = charByteSize;
+      // Treat invalid UTF-8 character as other char.
+      utf8proc_propval_t category =
+          curCodePoint == -1 ? 5 : utf8proc_category(curCodePoint);
```

Comment:
Would you clarify where the number 5 comes from? What is the problem if we keep using -1?

Reply:
5 means it is other letter as we treat invalid UTF-8 character now a other char
![image](https://github.com/user-attachments/assets/53abecde-e66b-4450-a8c7-4354492f82f3)

Reply:
I don't recognize why we cannot use the original value `-1` as isUpperChar/isLowerChar/isDigitChar check for their specific code and an invalid code cannot be matched.

Reply:
-1 is not a valid char category, but let me change to still use -1

Reply:
got your point, i will refactor to keep use original value. thanks.